server/repository/src/db_diesel/report_query.rs

```diff
@@ -0,0 +1,276 @@
+    let mut sql = if sql.chars().last() == Some(';') {
+        sql[..sql.len() - 1].to_string()
+    } else {
+        sql.to_string()
+    };
+    // Replace named variable like $myVariable with the numbered parameters like $1. Using the order
+    // in which variables where first used.
+    for (i, param) in used_params.iter().enumerate() {
+        sql = sql.replace(&param.0, &format!("${}", i + 1));
+    }
+    // Create the string containing all the parameter values
+    let param_values = used_params
+        .iter()
+        .map(|it| match &it.1 {
+            serde_json::Value::Null => "NULL".to_string(),
+            serde_json::Value::Bool(b) => format!("{b}"),
+            serde_json::Value::Number(n) => format!("{n}"),
+            serde_json::Value::String(s) => format!("'{}'", s),
+            // not supported but just add them...
+            serde_json::Value::Array(_) => format!("{}", it.1.to_string()),
+            serde_json::Value::Object(_) => format!("{}", it.1.to_string()),
+        })
+        .collect::<Vec<String>>()
+        .join(", ");
+    let param_values = if param_values.is_empty() {
+        "".to_string()
+    } else {
+        format!("({})", param_values)
+    };
+
+    // do the query
+    let pg_connection = &connection.connection;
+    let statement_name = format!("statement_{}", small_uuid());
+    let json_row_sql_query = format!(
+        "PREPARE {} AS
+            WITH provided_query AS(
+                {}
+                ) SELECT row_to_json(provided_query) as data FROM provided_query;
+        ",
+        statement_name, sql
+    );
+    pg_connection.batch_execute(&json_row_sql_query)?;
+    let json_results = sql_query(&format!("EXECUTE {}{};", statement_name, param_values))
+        .load::<JsonDataRow>(pg_connection)?;
+    pg_connection.batch_execute(&format!("DEALLOCATE PREPARE {};", statement_name))?;
+
+    let rows: Vec<serde_json::Value> = json_results
+        .into_iter()
+        .map(|r| serde_json::from_str(&r.data).unwrap())
+        .collect();
+
+    Ok(rows)
+}
+
+#[cfg(not(feature = "postgres"))]
+use crate::database_settings::DatabaseSettings;
+#[cfg(not(feature = "postgres"))]
+impl From<rusqlite::Error> for RepositoryError {
+    fn from(value: rusqlite::Error) -> Self {
+        RepositoryError::DBError {
+            msg: format!("{}", value),
+            extra: "".to_string(),
+        }
+    }
+}
+
+#[cfg(not(feature = "postgres"))]
+pub fn query_json(
+    settings: &DatabaseSettings,
+    sql: &str,
+    parameters: &serde_json::Map<String, serde_json::Value>,
+) -> Result<Vec<serde_json::Value>, RepositoryError> {
+    use rusqlite::{types::Null, Connection as RusqliteConnection};
+    use serde_json::Number;
+
+    let conn = RusqliteConnection::open(settings.connection_string())?;
+
+    let mut statement = conn.prepare(sql)?;
+
+    for p in 1..=statement.parameter_count() {
+        let Some(param) = statement.parameter_name(p) else {
```

Comment:
SQLite support named parameters, but I wonder if I should do the same manual SQL parsing as I have done for Postgres, just to have the same code/behaviour...